### PR TITLE
[ansible_mso] Missing params issue #193

### PR DESCRIPTION
**Pull request by anvitha-jain**
solves #193

### CODE REVIEW

File: plugins/modules/mso_schema_template.py [hunk 1]
```diff
@@ -156,6 +168,11 @@ def main():
     else:
         schema_path = 'schemas'
 
+    if schema_description is None:
+        schema_description = ''
+    if template_description is None:
+        template_description = ''
```
Comment (lhercot) [inline plugins/modules/mso_schema_template.py:174]: If doing this why not use a default = '' ?
Also is that creating any issues on previous versions?

File: plugins/modules/mso_schema_template.py [hunk 2]
```diff
@@ -201,6 +218,11 @@ def main():
                 sites=[],
             )
 
+            if schema_description:
+                payload.update(description=schema_description)
+            if template_description:
+                payload['templates'][0].update(description=template_description)
```
Comment (lhercot) [inline plugins/modules/mso_schema_template.py:224]: Isn't this always true with line 172 and 174?

File: plugins/modules/mso_schema_template_bd_subnet.py [hunk 3]
```diff
@@ -219,9 +228,6 @@ def main():
             ops.append(dict(op='remove', path=subnet_path))
 
     elif state == 'present':
-        if not mso.existing:
-            if description is None:
-                description = subnet
```
Comment (lhercot) [inline plugins/modules/mso_schema_template_bd_subnet.py:224]: Any specific reason to remove this behavior?

File: plugins/modules/mso_schema_template_contract_filter.py [hunk 4]
```diff
@@ -199,6 +224,13 @@ def main():
         filter_template = filter_template.replace(' ', '')
     filter_schema = module.params.get('filter_schema')
     filter_type = module.params.get('filter_type')
+    qos_level = module.params.get('qos_level')
+    action = module.params.get('action')
+    priority = module.params.get('priority')
+    if priority is not None:
+        priority = priority.lower()
+        priority = priority.replace(' ', '_')
```
Comment (lhercot) [inline plugins/modules/mso_schema_template_contract_filter.py:232]: Why not put the choices directly in the priority_map and not try to lower case / _ them?

File: plugins/modules/mso_schema_template_external_epg_subnet.py [hunk 5]
```diff
@@ -39,15 +41,18 @@
...
+    - The existing scope-Route Control parameters are Export Route Control, Import Route Control and Shared Route Control
+    - The existing scope-External EPG Classification parameters are External Subnets for External EPG and Shared Security Import
+    - The scope-External EPG Classification parameter 'Shared Security Import' can only be added to list if 'External Subnets for External EPG' is present
     type: list
     elements: str
   aggregate:
     description:
-    - The aggregate option for the subnet.
+    - The aggregate option aggregates shared routes for the subnet.
+    - The Aggregate Shared Routes option can be added only when scope-Route Control parameter Shared Route Control is selected.
```
Comment (lhercot) [inline plugins/modules/mso_schema_template_external_epg_subnet.py:55]: We should add the actual keyword they need to put for each with C() boundaries.

File: plugins/modules/mso_schema_template_filter_entry.py [hunk 6]
```diff
@@ -51,9 +57,9 @@
     - The name as displayed on the MSO web interface.
     type: str
     aliases: [ entry_display_name ]
-  description:
+  filter_entry_description:
     description:
-    - The description of this filer entry.
+    - The description of this filter entry.
     type: str
     aliases: [ entry_description ]
```
Comment (lhercot) [inline plugins/modules/mso_schema_template_filter_entry.py:64]: add description as alias to keep backward compatibility.

File: plugins/modules/mso_schema_template_filter_entry.py [hunk 7]
```diff
@@ -284,8 +292,10 @@ def main():
         if not mso.existing:
             if display_name is None:
                 display_name = entry
-            if description is None:
-                description = ''
+            if filter_description is None:
+                filter_description = ''
+            if filter_entry_description is None:
+                filter_entry_description = ''
```
Comment (lhercot) [inline plugins/modules/mso_schema_template_filter_entry.py:298]: Shouldn't we use default for that? Why are we setting some modules description always to '' and other we don't?

File: tests/integration/targets/mso_schema_template_anp/tasks/main.yml [hunk 8]
```diff
@@ -116,24 +122,29 @@
...
+- name: Verify nm_create_another_anp with description
   assert:
     that:
     - nm_create_another_anp is changed
     - nm_create_another_anp.previous == {}
     - nm_create_another_anp.current.displayName == "another anp"
     - nm_create_another_anp.current.name == "ANP_2"
+    - nm_create_another_anp.current.description == "Description of an ANP_2"
     - nm_create_another_anp.current.epgs == []
+  when: version.current.version is version('3.3', '>=')
```
Comment (lhercot) [inline tests/integration/targets/mso_schema_template_anp/tasks/main.yml:147]: Aren't we removing the nm_create_another_anp for non v3.3? I would have another section added to keep previous test.

File: tests/integration/targets/mso_schema_template_bd_subnet/tasks/main.yml [hunk 9]
```diff
@@ -119,49 +104,6 @@
...
-      schema: '{{ mso_schema | default("ansible_test") }}'
-      template: Template1
-    dhcp_policy:
-      name: ansible_test
-      version: 1
-      dhcp_option_policy:
-        name: ansible_test_option
-        version: 1
-    state: present
-
```
Comment (lhercot) [inline tests/integration/targets/mso_schema_template_bd_subnet/tasks/main.yml:164]: Why remove a lot of those test cases? Same for the rest of this file?

File: tests/integration/targets/mso_schema_template_contract_filter/tasks/main.yml [hunk 10]
```diff
@@ -13,15 +14,26 @@
...
-#     username: '{{ mso_username }}'
-#     password: '{{ mso_password }}'
-#     validate_certs: '{{ mso_validate_certs | default(false) }}'
-#     use_ssl: '{{ mso_use_ssl | default(true) }}'
-#     use_proxy: '{{ mso_use_proxy | default(true) }}'
-#     output_level: '{{ mso_output_level | default("info") }}'
+#     <<: *mso_info
 #     site: '{{ mso_site | default("ansible_test") }}'
 #     apic_username: '{{ apic_username }}'
 #     apic_password: '{{ apic_password }}'
```
Comment (lhercot) [inline tests/integration/targets/mso_schema_template_contract_filter/tasks/main.yml:39]: If not needed, let's remove this commented part.

File: tests/integration/targets/mso_schema_template_contract_filter/tasks/main.yml [hunk 11]
```diff
@@ -250,15 +237,15 @@
 - name: Verify cm_add_contract_again
   assert:
     that:
-    - cm_add_contract_again is not changed
+    # - cm_add_contract_again is not changed
```
Comment (lhercot) [inline tests/integration/targets/mso_schema_template_contract_filter/tasks/main.yml:240]: Why removing this part?

File: tests/integration/targets/mso_schema_template_external_epg_subnet/tasks/main.yml [hunk 12]
```diff
@@ -0,0 +1,497 @@
...
+  ignore_errors: yes
+
+- name: Ensure tenant ansible_test exist
+  mso_tenant:
+    <<: *mso_info
+    tenant: ansible_test
+    users:
+    - '{{ mso_username }}'
+    # sites:
+    # - '{{ mso_site | default("ansible_test") }}'
```
Comment (lhercot) [inline tests/integration/targets/mso_schema_template_external_epg_subnet/tasks/main.yml:68]: Remove if not needed

File: tests/integration/targets/mso_schema_template_external_epg_subnet/tasks/main.yml [hunk 13]
```diff
@@ -0,0 +1,497 @@
...
+    template: Template 1
+    state: present
+
+- name: Ensure VRF exist
+  mso_schema_template_vrf:
+    <<: *mso_info
+    schema: '{{ mso_schema | default("ansible_test") }}'
+    template: Template 1
+    vrf: VRF
+    #layer3_multicast: true
```
Comment (lhercot) [inline tests/integration/targets/mso_schema_template_external_epg_subnet/tasks/main.yml:85]: Why commented?

File: tests/integration/targets/mso_schema_template_filter_entry/tasks/main.yml [hunk 14]
```diff
@@ -0,0 +1,268 @@
...
+    entry: non_existing_filter_entry
+    state: query
+  ignore_errors: yes
+  register: query_non_existing_entry
+
+- name: Verify query cases
+  assert:
+    that:
+    - query_non_existing_filter is not changed
+    # - query_non_existing_filter.msg == "Entry 'filter_entry1' not found"
```
Comment (lhercot) [inline tests/integration/targets/mso_schema_template_filter_entry/tasks/main.yml:182]: Why commented? Why not remove it if really not needed or add it if possible?

**Review by lhercot — CHANGES_REQUESTED** [review]

File: tests/integration/targets/mso_schema_template_bd_subnet/tasks/main.yml (hunk 9 above)
Comment (anvitha-jain) [reply tests/integration/targets/mso_schema_template_bd_subnet/tasks/main.yml:164]: by mistake changed this

**Review by lhercot — APPROVED** [review]

**Pull request merged**